runtime/common/JsonConvert.h

```diff
@@ -178,10 +197,236 @@ inline void from_json(const json &j, ExecutionContext &context) {
+  TO_JSON_OPT_HELPER(step_size);
+}
+
+inline void to_json(json &j, const cudaq::optimizers::base_nlopt &p) {
+  TO_JSON_OPT_HELPER(max_eval);
+  TO_JSON_OPT_HELPER(initial_parameters);
+  TO_JSON_OPT_HELPER(lower_bounds);
+  TO_JSON_OPT_HELPER(upper_bounds);
+  TO_JSON_OPT_HELPER(f_tol);
+}
+
+inline void to_json(json &j, const cudaq::optimizer &p) {
+  if (auto *base_ensmallen =
+          dynamic_cast<const cudaq::optimizers::BaseEnsmallen *>(&p))
+    j = json(*base_ensmallen);
+  else if (auto *base_nlopt =
+               dynamic_cast<const cudaq::optimizers::base_nlopt *>(&p))
+    j = json(*base_nlopt);
+}
+
+inline void from_json(const nlohmann::json &j,
+                      cudaq::optimizers::BaseEnsmallen &p) {
+  FROM_JSON_OPT_HELPER(max_eval);
+  FROM_JSON_OPT_HELPER(initial_parameters);
+  FROM_JSON_OPT_HELPER(lower_bounds);
+  FROM_JSON_OPT_HELPER(upper_bounds);
+  FROM_JSON_OPT_HELPER(f_tol);
+  FROM_JSON_OPT_HELPER(step_size);
+}
+
+inline void from_json(const nlohmann::json &j,
+                      cudaq::optimizers::base_nlopt &p) {
+  FROM_JSON_OPT_HELPER(max_eval);
+  FROM_JSON_OPT_HELPER(initial_parameters);
+  FROM_JSON_OPT_HELPER(lower_bounds);
+  FROM_JSON_OPT_HELPER(upper_bounds);
+  FROM_JSON_OPT_HELPER(f_tol);
+}
+
+inline std::unique_ptr<cudaq::optimizer>
+make_optimizer_from_json(const nlohmann::json &j,
+                         const OptimizerEnum optimizer_type) {
+  if (optimizer_type == OptimizerEnum::COBYLA) {
+    auto ret_ptr = std::make_unique<cudaq::optimizers::cobyla>();
+    from_json(j, dynamic_cast<cudaq::optimizers::base_nlopt &>(*ret_ptr));
+    return ret_ptr;
+  }
+  if (optimizer_type == OptimizerEnum::NELDERMEAD) {
+    auto ret_ptr = std::make_unique<cudaq::optimizers::neldermead>();
+    from_json(j, dynamic_cast<cudaq::optimizers::base_nlopt &>(*ret_ptr));
+    return ret_ptr;
+  }
+  if (optimizer_type == OptimizerEnum::LBFGS) {
+    auto ret_ptr = std::make_unique<cudaq::optimizers::lbfgs>();
+    from_json(j, dynamic_cast<cudaq::optimizers::BaseEnsmallen &>(*ret_ptr));
+    return ret_ptr;
+  }
+  if (optimizer_type == OptimizerEnum::SPSA) {
+    auto ret_ptr = std::make_unique<cudaq::optimizers::spsa>();
+    from_json(j, dynamic_cast<cudaq::optimizers::BaseEnsmallen &>(*ret_ptr));
+    return ret_ptr;
+  }
+  if (optimizer_type == OptimizerEnum::ADAM) {
+    auto ret_ptr = std::make_unique<cudaq::optimizers::adam>();
+    from_json(j, dynamic_cast<cudaq::optimizers::BaseEnsmallen &>(*ret_ptr));
+    return ret_ptr;
+  }
+  if (optimizer_type == OptimizerEnum::GRAD_DESC) {
+    auto ret_ptr = std::make_unique<cudaq::optimizers::gradient_descent>();
+    from_json(j, dynamic_cast<cudaq::optimizers::BaseEnsmallen &>(*ret_ptr));
+    return ret_ptr;
+  }
+  if (optimizer_type == OptimizerEnum::SGD) {
+    auto ret_ptr = std::make_unique<cudaq::optimizers::sgd>();
+    from_json(j, dynamic_cast<cudaq::optimizers::BaseEnsmallen &>(*ret_ptr));
+    return ret_ptr;
+  }
```

Comment:
nit: is there a reason we don't use `switch case` here?

Reply:
Updated in e8a16e4db. I also removed the unnecessary dynamic_cast just to tidy it up.